STM32F1/libraries/EEPROM/EEPROM.cpp

```diff
@@ -212,56 +212,71 @@ uint16 EEPROMClass::EE_VerifyPageFullWriteVariable(uint16 Address, uint16 Data)
 	FLASH_Status FlashStatus;
 	uint32 idx, pageBase, pageEnd, newPage;
 	uint16 count;
-
+	boolean found = false;
 	// Get valid Page for write operation
+	
 	pageBase = EE_FindValidPage();
 	if (pageBase == 0)
 		return  EEPROM_NO_VALID_PAGE;
-
+	
 	// Get the valid Page end Address
 	pageEnd = pageBase + PageSize;			// Set end of page
-
-	for (idx = pageEnd - 2; idx > pageBase; idx -= 4)
+	
+	//FLASH_Unlock();
+	
+	for (idx = pageEnd - 2; idx >= pageBase; idx -= 4)
 	{
 		if ((*(__IO uint16*)idx) == Address)		// Find last value for address
 		{
+			found = false;
+			
 			count = (*(__IO uint16*)(idx - 2));	// Read last data
+			
 			if (count == Data)
 				return EEPROM_OK;
-			if (count == 0xFFFF)
+			
+			if (count == 0xFFFF || Data < count)
```

Comment:
This is incorrect, you can't reprogram 1s in a halfword of flash, only 0s, so comparing Data < count  to decide whether the word can be rewritten is invalid and will lead to corruption.
Either leave is as before (comparing to 0xFFFF) or compare to 0 to make sure the new data has all bits 0.
As per the flash programming manual:

> Standard programming
In this mode the CPU programs the main Flash memory by performing standard half-word
write operations. The PG bit in the FLASH_CR register must be set. FPEC preliminarily
reads the value at the addressed main Flash memory location and checks that it has been
erased. If not, the program operation is skipped and a warning is issued by the PGERR bit in
FLASH_SR register (the only exception to this is when 0x0000 is programmed. In this case,
the location is correctly programmed to 0x0000 and the PGERR bit is not set).

`if (count == 0xFFFF || Data = 0)`